File: Emisor/Emisor.java
class ComputeHammingCode {

    public static String computeHammingCode(String data) {
        int r = calculateRedundantBits(data.length());
        String arr = addRedundantBits(data, r);
        arr = calculateParityBits(arr, r);
        return arr;
    }

    public static int calculateRedundantBits(int m) {
        for (int i = 0; i < m; i++) {
            if (Math.pow(2, i) >= m + i + 1) {
                return i;
            }
        }
        return 0;
    }

    public static String addRedundantBits(String data, int r) {
        int j = 0;
        int k = 1;
        int m = data.length();
        StringBuilder res = new StringBuilder();

        for (int i = 1; i <= m + r; i++) {
            if (i == Math.pow(2, j)) {
                res.append('0');
                j++;
            } else {
                res.append(data.charAt(m - k));
                k++;
            }
        }

        return res.reverse().toString();
    }

    public static String calculateParityBits(String arr, int r) {
        int n = arr.length();
        StringBuilder result = new StringBuilder(arr);

        for (int i = 0; i < r; i++) {
            int val = 0;
            for (int j = 1; j <= n; j++) {
                if ((j & (1 << i)) == (1 << i)) {
                    val ^= Integer.parseInt(String.valueOf(arr.charAt(n - j)));
                }
            }
            result.setCharAt(n - (1 << i), (char) (val + '0'));
        }

        return result.toString();
    }

    public static void main(String[] args) {
        String data = "1011001";
        String hammingCode = computeHammingCode(data);
        System.out.println("Transferred data with Hamming code: " + hammingCode);
    }
}

File: Emisor/Emisor.py
def compute_hamming_code(data):
    def calculate_redundant_bits(m):
        for i in range(m):
            if 2 ** i >= m + i + 1:
                return i

    def add_redundant_bits(data, r):
        j = 0
        k = 1
        m = len(data)
        res = ''

        for i in range(1, m + r + 1):
            if i == 2 ** j:
                res = res + '0'
                j += 1
            else:
                res = res + data[-1 * k]
                k += 1

        return res[::-1]

    def calculate_parity_bits(arr, r):
        n = len(arr)

        for i in range(r):
            val = 0
            for j in range(1, n + 1):
                if j & (2 ** i) == (2 ** i):
                    val = val ^ int(arr[-1 * j])
            arr = arr[:n - (2 ** i)] + str(val) + arr[n - (2 ** i) + 1:]

        return arr

    r = calculate_redundant_bits(len(data))
    arr = add_redundant_bits(data, r)
    arr = calculate_parity_bits(arr, r)
    return arr

data = '110011011110101010101'
print("Data transferred is " + data)
hamming_code = compute_hamming_code(data)
print("Transferred data with Hamming code: " + hamming_code)

